ptvsd/_remote.py

```diff
@@ -28,63 +31,51 @@ def _pydevd_settrace(redirect_output=None, _pydevd=pydevd, **kwargs):
 # Then move at least some parts to the appropriate modules.  This module
 # is focused on running the debugger.
 
-def enable_attach(address,
-                  on_attach=(lambda: None),
-                  redirect_output=True,
-                  _pydevd=pydevd,
-                  _install=install,
-                  _settrace=_pydevd_settrace,
-                  **kwargs):
-    addr = Address.as_server(*address)
-    debug('installing ptvsd as server')
-    # pydevd.settrace() forces a "client" connection, so we trick it
-    # by setting start_client to start_server..
-    daemon = _install(
-        _pydevd,
-        addr,
-        start_client=start_server,
-        notify_session_debugger_ready=(lambda s: on_attach()),
-        singlesession=False,
-        **kwargs
-    )
-
-    def start_pydevd():
-        debug('enabling pydevd')
-        # Only pass the port so start_server() gets triggered.
-        # As noted above, we also have to trick settrace() because it
-        # *always* forces a client connection.
-        _settrace(
-            host=addr.host,
-            stdoutToServer=redirect_output,
-            stderrToServer=redirect_output,
-            port=addr.port,
-            suspend=False,
-            _pydevd=_pydevd,
-        )
-        debug('pydevd enabled')
-    t = new_hidden_thread('start-pydevd', start_pydevd)
-    t.start()
-
-    def wait(timeout=None):
-        t.join(timeout)
-        return not t.is_alive()
-
-    def debug_current_thread(suspend=False, **kwargs):
-        # Make sure that pydevd has finished starting before enabling
-        # in the current thread.
-        t.join()
-        debug('enabling pydevd (current thread)')
-        _settrace(
-            host=None,  # ignored
-            stdoutToServer=False,  # ignored
-            stderrToServer=False,  # ignored
-            port=None,  # ignored
-            suspend=suspend,
-            trace_only_current_thread=True,
-            overwrite_prev_trace=True,
-            patch_multiprocessing=False,
-            _pydevd=_pydevd,
-            **kwargs
-        )
-        debug('pydevd enabled (current thread)')
-    return daemon, wait, debug_current_thread
+global_next_session = None
+
+
+def enable_attach(address, redirect_output=True,
+                  _pydevd=pydevd, _install=install,
+                  on_attach=lambda: None, **kwargs):
+    host, port = address
+
+    def wait_for_connection(daemon, host, port, next_session=None):
+        debugger = get_global_debugger()
+        while debugger is None:
+            time.sleep(0.1)
+            debugger = get_global_debugger()
+
+        debugger.ready_to_run = True
+
+        while True:
+            session_not_bound.wait()
+            try:
+                global_next_session()
+                on_attach()
+            except DaemonClosedError:
+                return
+
+    def start_daemon():
+        daemon._sock = daemon._start()
```

Comment:
>daemon._sock = daemon._start() [](start = 8, length = 30)

Why is this being done outside of `_start()`?

Reply:
After offline discussion. This is fine for now.

---
In reply to: [206973751](https://github.com/Microsoft/ptvsd/pull/706#discussion_r206973751) [](ancestors = 206973751)

Reply:
> Why is this being done outside of _start()?

Simple, time constraints. I could either leave the hack, or spend a month fixing it :)